proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,510 @@
+
+```haskell
+data T = T -- with a single unified namespace, T the data would have a name clash with T the type.
+
+a = 15
+
+f :: forall a. a -> a -- no warning, because with a single unified namespace 'forall a.' would shadow the top level 'a', so 'a' in 'a -> a' would still refer to the forall bound 'a'.
+```
+
+There are more detailed examples in Examples section.
+
+## Proposed Change Specification
+
+* To disambiguate the namespaces of identifiers from modules that use punning,
+  the `-XExplicitNamespaces` extension is extended with new syntax:
+
+  ```
+  impdecl   -> import [qualified] modid [as modid] [impspec]
+  ```
+  is changed to
+  ```
+  impdecl   -> import [qualified] modid [namespace] [as modid] [impspec]
+
+  namespace -> data
+             | type
+  ```
+   * With `data` specified in the import, only identifiers belonging to the data namespace will be brought into the scope.
+   * With `type` specified in the import, only identifiers belonging to the type namespace will be brought into the scope.
+
+* To disambiguate the namespaces within one module, a `-XModuleAliases` extension
+  is added that introduces new syntax:
+
+  ```
+  module   -> module modid [exports] where body
+            | body
+  ```
+  is changed to
+  ```
+  module   -> module modid [alias_1, ..., alias_n] [exports] where body
+            | body
+
+  alias -> [namespace] as modid
+
+  namespace -> data
+             | type
+  ```
+    * When `as modid` is specified, the users can refer to the module within
+      the module itself using the provided qualifier.
+    * When `data as modid` is specified, the users can refer to the terms
+      in the module within the module itself using the provided qualifier.
+    * When `type as modid` is specified, the users can refer to the types
+      in the module within the module itself using the provided qualifier.
+
+* Deprecate the `'` syntax of `-XDataKinds`, reserving this syntax for Template
+  Haskell name quotation. Introduce a new warning, `-Wticked-promoted-constructors` which
+  warns of the usage of the syntax.
+
+* Deprecate the `''` syntax in Template Haskell. Introduce a new warning,
+  `-Wdouble-tick-template-haskell` which warns about the usage of the syntax.
+
+* Introduce a new warning, `-Wpun-bindings`, triggered by any name
+  binding that would clash with another identifier if Haskell had a single
+  unified namespace.
+
+* Introduce a new warning, `-Wpuns`, triggered by using an identifier
+  that would be ambiguous or refer to another entity if Haskell had a single
+  unified namespace.
+
+* The deprecation strategy for the `'` syntax in `-XDataKinds` and `''` syntax in `-XTemplateHaskell` is the following:
+  * In the next release add `-Wpun-bindings`, `-Wpuns`, `-Wticked-promoted-constructors` and `-Wdouble-tick-template-haskell` to `-Wcompat` and deprecate `-Wunticked-promoted-constructors`.
+  * Three releases from after this proposal is implemented add all four warnings to `-Wall`.
```

Comment:
This is major, ecosystem-sized breakage, given the prevalence of `-Wall`. I think we'll need a large community discussion before blazing ahead with putting these in `-Wall`.